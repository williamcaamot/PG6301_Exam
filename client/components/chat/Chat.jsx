import React, {createContext, useEffect, useState} from "react";
import ChatRoomListing from "./ChatRoomListing.jsx";
import ChatWindow from "./ChatWindow.jsx";
import ErrorMessage from "../globals/ErrorMessage.jsx";


export const ChatContext = createContext();

export function Chat() {
    const [errorMessage, setErrorMessage] = useState();

    const [ws, setWs] = useState(null);
    const [messages, setMessages] = useState([]);
    const [chatRooms, setChatRooms] = useState([]);
    const [activeChatRoom, setActiveChatRoom] = useState(null);
    const [activeChatRoomTitle, setActiveChatRoomTitle] = useState()


    async function handleFetchChatrooms() {
        try {
            const res = await fetch("/api/v1/chatroom");
            const {message, data} = await res.json();
            if (res.status !== 200) {
                setErrorMessage(message);
                return;
            }
            setChatRooms(data);

        } catch (e) {
            setErrorMessage(e.message);
        }
    }
    useEffect(() => { //Remove any exisiting listeneres and sockets before establishing a new one on active chat room change
        if (ws) {
            ws.onmessage = null;
            ws.close();
        }
    }, [activeChatRoom]);

    useEffect(() => {
        //Try to establish socket here and put it in the context
        const ws = new WebSocket(window.location.origin.replace(/^http/, "ws"));
        setWs(ws);
        ws.onmessage = (event) => {
            const {chatroomid, message, errormessage} = JSON.parse(event.data);
            if(errormessage){
                setErrorMessage(errormessage);
                return;
            }
            if (chatroomid === activeChatRoom) {
                console.log("Updating messages from socket")
                setMessages(prevMessages => [...prevMessages, message]);
            }
        }
    }, [activeChatRoom]);



    async function setChatroom(id, title) {
        setActiveChatRoom(id);
        setActiveChatRoomTitle(title);
    }

    useEffect(() => {
        handleFetchChatrooms();
    }, []);

    


    return <ChatContext.Provider value={{
        chatRooms,
        activeChatRoom,
        ws,
        messages,
        setMessages,
        activeChatRoomTitle,
        setChatroom
    }}>
        <div className={"pageContentWrapper"}>
            <div className={"innerWrapper"}>
                <ErrorMessage message={errorMessage}/>
                <div style={{width: "100%", display: "flex"}}>
                    <div className={"chatRooms"}>
                        <div style={{width: "100%", display: "flex", flexWrap: "nowrap"}}>
                            <h2>Available chatrooms:</h2>
                        </div>
                        <div style={{width: "100%"}}>
                            {chatRooms && chatRooms.map(e => {
                                return <ChatRoomListing
                                    chatRoom={e}
                                    key={e._id}
                                />
                            })}
                        </div>
                    </div>


                    <div className={"chatWindow"} style={{padding: "15px"}}>

                        {activeChatRoom && <ChatWindow/>}
                    </div>


                </div>
            </div>
        </div>
    </ChatContext.Provider>
}