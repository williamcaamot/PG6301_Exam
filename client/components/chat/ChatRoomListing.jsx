import React, {useContext} from "react";
import {ChatContext} from "./Chat.jsx";


function ChatRoomListing({chatRoom}) {

    const {setChatroom} = useContext(ChatContext);

    return <>
        <div style={{paddingTop: "15px", width:"100%"}}>
            <div className={"chatRoomListing"}
                onClick={(e) => setChatroom(chatRoom._id, chatRoom.title)}>
                <h3>{chatRoom.title}</h3>
                <p>{chatRoom.description}</p>
            </div>
        </div>
    </>
}

export default ChatRoomListing;