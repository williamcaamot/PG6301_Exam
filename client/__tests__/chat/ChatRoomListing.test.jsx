import renderer, {act} from "react-test-renderer";
import {MemoryRouter} from "react-router-dom";
import ChatRoomListing from "../../components/chat/ChatRoomListing.jsx";
import {createContext} from "react";
import {ChatContext} from "../../components/chat/Chat.jsx";

describe("ChatRoomListing", () => {

    it("should render a chat room", () => {
        async function setChatroom(){
        }

        const user = {
            _id: "655dd38a4c15f00c20bc3fcf",
            bio: "dawawddawJeg er kul",
            email: "williamcoucheronaamot@gmail.com",
            family_name: "Coucheron-Aamot",
            name: "William",
            nickname: "WILLIAM123wwwww",
            picture: "https://lh3.googleusercontent.com/a/ACg8ocIOgSmXkWTJqoKPe5Hk3JlkkhfUNer4sWkVo3UuIa_dzg=s96-c",
        }

        const chatroom = {
            _id: "655e00a9f595fb7c60770caa",
            description: "dawadwadwadw",
            owner: "williamcoucheronaamot@gmail.com",
            title: "dawdwaadw"
        }

        const component2 = renderer.create(
            <ChatContext.Provider value={setChatroom}>
                <MemoryRouter>
                    <ChatRoomListing chatRoom={{chatroom}}/>
                </MemoryRouter>)
            </ChatContext.Provider>
        )
        expect(component2).toMatchSnapshot();
    })


})