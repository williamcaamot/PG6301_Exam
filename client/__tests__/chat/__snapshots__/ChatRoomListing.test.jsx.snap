// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatRoomListing should render a chat room 1`] = `
[
  <div
    style={
      {
        "paddingTop": "15px",
        "width": "100%",
      }
    }
  >
    <div
      className="chatRoomListing"
      onClick={[Function]}
    >
      <h3 />
      <p />
    </div>
  </div>,
  ")",
]
`;
